### PR TITLE
add port queue stats handler API

### PR DESCRIPTION
**Pull request by archerhungbsn**
Reviewer: @kenchiang @nick-bsn 
Due to port queue stats buffer overrun in wildcard case, the new API will handle the multiple reply messages.

### CODE REVIEW

File: modules/OFStateManager/module/src/handlers.c [hunk 1]
```diff
@@ -312,6 +312,18 @@ ind_core_queue_stats_request_handler(of_object_t *_obj, indigo_cxn_id_t cxn_id)
 
         of_queue_stats_entry_delete(queue_stats);
         indigo_cxn_send_controller_message(cxn_id, reply);
+    } else if (indigo_port_queue_stats_get_handler) {
+        rv = indigo_port_queue_stats_get_handler(obj, cxn_id);
+        if (rv == INDIGO_ERROR_NONE) {
```
Comment (kenchiang) [inline modules/OFStateManager/module/src/handlers.c:317]: it looks like we're calling indigo_port_queue_stats_get_handler if that function is defined.
but if that function returns INDIGO_ERROR_NONE, then we're going to send an error reply?
maybe i read this code too quickly and missed something obvious.

File: modules/indigo/module/inc/indigo/port_manager.h [hunk 2]
```diff
@@ -225,6 +225,22 @@ extern indigo_error_t indigo_port_queue_config_get(
     of_queue_get_config_reply_t **queue_config_reply);
 
 
+/**
+ * @brief Process an OF queue stats request handler
```
Comment (kenchiang) [inline modules/indigo/module/inc/indigo/port_manager.h:229]: Maybe reword this to: "Prototype of an OF queue stats request handler; if defined, this handler will be called when a queue stats request is received"

File: modules/indigo/module/inc/indigo/port_manager.h [hunk 3]
```diff
@@ -225,6 +225,22 @@ extern indigo_error_t indigo_port_queue_config_get(
...
+ * @param queue_stats_request The LOXI request message
+ * @param cxn_id Connection ID on which the message arrived
+ * @return Return code from operation
```
Comment (kenchiang) [inline modules/indigo/module/inc/indigo/port_manager.h:232]: Should there be some english such as "The handler should return INDIGO_ERROR_NONE if the request has been successfully handled; otherwise an error reply will be sent with type XXX and code YYY" ?

File: modules/OFStateManager/module/src/handlers.c [hunk 4]
```diff
@@ -312,6 +312,18 @@ ind_core_queue_stats_request_handler(of_object_t *_obj, indigo_cxn_id_t cxn_id)
...
+            of_port_no_t port_no;
+            uint32_t queue_id;
+            of_queue_stats_request_port_no_get(obj, &port_no);
+            of_queue_stats_request_queue_id_get(obj, &queue_id);
+            AIM_LOG_ERROR("Failed to get stats for queue %u on port %u: %s",
+                          queue_id, port_no, indigo_strerror(rv));
+            /* @todo sending type 0, code 0 error message */
+            indigo_cxn_send_error_reply(cxn_id, obj, 0, 0);
```
Comment (kenchiang) [inline modules/OFStateManager/module/src/handlers.c:325]: are there more applicable type and code (other than 0)?

**Review by kenchiang — COMMENTED** [review]
This looks like a good approach to address this problem.
I'm wondering if OFStateManager just shouldn't do anything if `indigo_port_queue_stats_get_handler` returns INDIGO_ERROR_NONE.

File: modules/OFStateManager/module/src/handlers.c (hunk 1 above)
Comment (archerhungbsn) [reply modules/OFStateManager/module/src/handlers.c:317]: Originally, I planned not to pass the cxn_id to the handler() so that the error status is handled in indigo. If handler() function is ok to return the error message, then indigo can do nothing when error is returned.

Comment (archerhungbsn) [reply modules/OFStateManager/module/src/handlers.c:317]: I will let the handler() to return the error as suggested

File: modules/indigo/module/inc/indigo/port_manager.h (hunk 3 above)
Comment (archerhungbsn) [reply modules/indigo/module/inc/indigo/port_manager.h:232]: Will let the handler() reply the error code to the controller. The return code here will be casted to void and ignored by indigo.

File: modules/OFStateManager/module/src/handlers.c (hunk 1 above)
Comment (kenchiang) [reply modules/OFStateManager/module/src/handlers.c:317]: i think we should provide some more detail for the comment that precedes indigo_port_queue_stats_get_handler in port_manager.h.  if the handler is supposed to deal with any error, then the prototype of the handler should return void, and the comment should say "handler should send reply or error".

alternately, you could keep the handler prototype as it is, and your code in handler.c can send an error with appropriate type and code if the rv != INDIGO_ERROR_NONE.  so if the driver wants to send an error, it can do so inside the handler, and the handler can return INDIGO_ERROR_NONE so that no error is sent by OFStateManager.

so just pick one or the other, but from your comment above, it sounds like you're going with the first option;  can you confirm?

File: modules/indigo/module/inc/indigo/port_manager.h (hunk 3 above)
Comment (kenchiang) [reply modules/indigo/module/inc/indigo/port_manager.h:232]: ok, if the handler returns the error code to the controller, then the prototype for the handler should have "void" as its return type.

**Comment by kenchiang** [comment]
LGTM, thanks for taking care of this.
@nick-bsn : what do you think?

**Comment by nick-bsn** [comment]
@kenchiang LGTM too. 👍 Thanks a lot for the review!
